lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp

```diff
@@ -3088,105 +3178,140 @@ LogicalResult CppEmitter::emitOperation(Operation &op, bool trailingSemicolon) {
-  if (auto iType = type.dyn_cast<IndexType>())
-    return (os << "size_t"), success();
+  if (auto iType = dyn_cast<IndexType>(type))
+    return "size_t";
 
-  if (auto tType = type.dyn_cast<TensorType>()) {
+  if (auto tType = dyn_cast<TensorType>(type)) {
     if (!tType.hasRank())
-      return emitError(loc, "cannot emit unranked tensor type");
+      return {};
     if (!tType.hasStaticShape())
-      return emitError(loc, "cannot emit tensor type with non static shape");
-    os << "Tensor<";
-    if (failed(emitType(loc, tType.getElementType())))
-      return failure();
+      return {};
+    ss << "Tensor<";
+    auto nestedTypeName = genCppTypeName(tType.getElementType());
+    if (!nestedTypeName)
+      return {};
+    ss << *nestedTypeName;
     auto shape = tType.getShape();
     for (auto dimSize : shape) {
-      os << ", ";
-      os << dimSize;
+      ss << ", ";
+      ss << dimSize;
     }
-    os << ">";
-    return success();
-  }
-  if (auto tType = type.dyn_cast<TupleType>())
-    return emitTupleType(loc, tType.getTypes());
-  if (auto oType = type.dyn_cast<emitc::OpaqueType>()) {
-    os << oType.getValue();
-    return success();
+    ss << ">";
+    return ss.str();
+  }
+  if (auto tType = dyn_cast<TupleType>(type)) {
+    ss << "std::tuple<";
+    bool itrleaveFailed = false;
+    llvm::interleave(
+        tType.getTypes(),
+        [&](Type type) {
+          auto optTyNameStr = genCppTypeName(type);
+          if (optTyNameStr)
+            ss << *optTyNameStr;
+          else
+            itrleaveFailed = true;
+        },
+        [&]() { ss << ", "; });
+    ss << ">";
+    if (!itrleaveFailed)
+      return ss.str();
+    return {};
+  }
+  if (auto oType = dyn_cast<emitc::OpaqueType>(type)) {
+    ss << oType.getValue().str();
+    return ss.str();
   }
   // Types added for AIE
   // MemRefType: printed as 'eltType'*
-  if (auto tType = type.dyn_cast<MemRefType>()) {
-    if (failed(emitType(loc, tType.getElementType())))
-      return failure();
-    os << " * restrict";
-    return success();
+  if (auto tType = dyn_cast<MemRefType>(type)) {
+    auto elemTyStrOpt = genCppTypeName(tType.getElementType());
+    if (!elemTyStrOpt)
+      return {};
+    ss << *elemTyStrOpt << " * restrict";
+    return ss.str();
   }
   // VectorType: printed as v'lane''eltType'
-  if (auto tType = type.dyn_cast<VectorType>()) {
+  if (auto tType = dyn_cast<VectorType>(type)) {
     Type eltType = tType.getElementType();
-    if (tType.getRank() != 1)
-      return failure();
-
-    unsigned dimSize = tType.getDimSize(tType.getRank() - 1);
-    os << "v" << std::to_string(dimSize);
+    // Flatten multidimensional vectors
+    auto vShape = tType.getShape();
+    int64_t numElems = std::accumulate(vShape.begin(), vShape.end(), 1,
+                                       std::multiplies<int64_t>());
+    ss << "v" << std::to_string(numElems);
+    if (numElems * tType.getElementTypeBitWidth() == 1024)
+      isAcc = true;
```

Comment:
I want to discuss this before merging. It's convenient because it automatically deducts a 1024-bit parameter to a function that takes vector types, must be an accumulator. The whole C++ translation is full of hacks and choppy design, it wouldn't be entirely out of place, but I want to consider better options with you lot.